packages/react/src/useActionForm.ts

```diff
@@ -0,0 +1,263 @@
+     */
+    onError?: (error: Error | FieldErrors<ActionFunc["variablesType"]>) => void;
+  }
+): UseActionFormResult<GivenOptions, SchemaT, ActionFunc, FormContext> => {
+  const api = useApi();
+  const findExistingRecord = !!options?.findBy;
+  const hasSetInitialValues = useRef<boolean>(!findExistingRecord);
+
+  // find the existing record if there is one
+  const modelManager = (api as any)[action.modelApiIdentifier] as AnyModelManager;
+  const [findResult] = useFindExistingRecord(modelManager, options?.findBy || "1", {
+    pause: !findExistingRecord,
+    select: options?.select,
+  });
+
+  const existingRecord = options?.record || findResult.data;
+  const isReady = !findExistingRecord || !!existingRecord;
+
+  let defaultValues = options?.defaultValues;
+  if (!defaultValues && existingRecord) {
+    defaultValues = { [action.modelApiIdentifier]: { ...existingRecord } } as any;
+  }
+
+  // setup the react-hook-form object, passing any options from the props
+  const { handleSubmit, formState, ...formHook } = useForm<ActionFunc["variablesType"], FormContext>({
+    defaultValues,
+    ...options,
+  });
+
+  // when the default values arrive from the record find later, reset them into the form. react-hook-form doesn't watch the default values after the first render
+  useEffect(() => {
+    if (isReady && !hasSetInitialValues.current && defaultValues) {
+      hasSetInitialValues.current = true;
+      formHook.reset(defaultValues);
+    }
+  }, [isReady, defaultValues, formHook]);
+
+  // get the action runner to run on submit
+  const [actionResult, runAction] = useAction(action, { select: options?.select } as any);
+
+  const handleSubmissionError = useCallback(
+    (error: Error | FieldErrors<ActionFunc["variablesType"]>) => {
+      if ("executionErrors" in error) {
+        const executionErrors = (error as unknown as ErrorWrapper).executionErrors;
+        for (const executionError of executionErrors) {
+          if ("validationErrors" in executionError) {
+            for (const validationError of executionError.validationErrors) {
+              formHook.setError(`${action.modelApiIdentifier}.${validationError.apiIdentifier}` as any, {
+                message: validationError.message,
+              });
+            }
+          } else {
+            const codeToReplace = "code" in executionError ? `${executionError.code}: ` : "";
+            const message = executionError.message.replace(codeToReplace, "");
+
+            formHook.setError("root", { message });
+          }
+        }
+      }
+
+      options?.onError?.(error);
+    },
+    [action.modelApiIdentifier, formHook, options]
+  );
+
+  const submit = useCallback(
+    async (event?: React.BaseSyntheticEvent<object, any, any> | undefined) => {
+      let result: any;
+
+      formHook.clearErrors();
+
+      await handleSubmit(
+        async (data) => {
+          let modelInput = data[action.modelApiIdentifier];
```

Comment:
Hard to tell but how does this case arise? How do you set properties on the data such that the inputs are set on the model api identifier? I guess how would you use `useActionForm` with a model that has ambiguous identifiers.

Reply:
added tests for this 👍